pkg/util/lock.go

```diff
@@ -111,30 +114,167 @@ func (l *PgAdvisoryLock) TryLock() (bool, error) {
+	if maxRetries == cur {
+		return nil, fmt.Errorf("max attempts reached. giving up on getting a db connection")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), waitForConnectionTimeout)
+	defer cancel()
+
+	cfg, err := pgx.ParseConfig(connStr)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing config connection: %w", err)
+	}
+
+	lockConn, err := pgx.ConnectConfig(ctx, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("error getting DB connection: %w", err)
 	}
-	defer func() {
+
+	if l.afterConnect != nil {
+		err = l.afterConnect(ctx, lockConn)
 		if err != nil {
-			l.connCleanUp()
+			log.Error("msg", "Invalid connection", "err", err)
+			return l.getConn(connStr, cur+1, maxRetries)
 		}
-	}()
-	rows, err := l.conn.Query(context.Background(), "SELECT pg_try_advisory_lock($1)", l.groupLockID)
+	}
+
+	return lockConn, nil
+}
+
+func (l *PgAdvisoryLock) GetAdvisoryLock() (bool, error) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	return l.getAdvisoryLock()
+}
+
+func (l *PgAdvisoryLock) getAdvisoryLock() (bool, error) {
+	return l.runLockFuncion("SELECT pg_try_advisory_lock($1)")
+}
+
+func (l *PgAdvisoryLock) GetSharedAdvisoryLock() (bool, error) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	return l.getSharedAdvisoryLock()
+}
+
+func (l *PgAdvisoryLock) getSharedAdvisoryLock() (bool, error) {
+	return l.runLockFuncion("SELECT pg_try_advisory_lock_shared($1)")
+}
+
+func (l *PgAdvisoryLock) runLockFuncion(query string) (bool, error) {
+	err := l.ensureConnInit()
+	if err != nil {
+		return false, err
+	}
+	return runLockFuncion(context.Background(), l.conn, query, l.groupLockID)
+}
+
+func (l *PgAdvisoryLock) ensureConnInit() error {
+	if l.conn != nil {
+		return nil
+	}
+	conn, err := l.getConn(l.connStr, 0, 10)
+	l.conn = conn
+	return err
+}
+
+func (l *PgAdvisoryLock) Unlock() (bool, error) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	return l.unlock()
+}
+
+func (l *PgAdvisoryLock) unlock() (bool, error) {
+	return l.runLockFuncion("SELECT pg_advisory_unlock($1)")
+}
+
+func (l *PgAdvisoryLock) UnlockShared() (bool, error) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	return l.unlockShared()
+}
+
+func (l *PgAdvisoryLock) unlockShared() (bool, error) {
+	return l.runLockFuncion("SELECT SELECT pg_advisory_unlock_shared($1)")
+}
+
+func runLockFuncion(ctx context.Context, conn *pgx.Conn, query string, lockId int64) (bool, error) {
+	rows, err := conn.Query(ctx, query, lockId)
 	if err != nil {
 		return false, err
 	}
 	defer rows.Close()
 	if !rows.Next() {
 		err = rows.Err()
 		if err != nil {
-			return false, fmt.Errorf("error while trying to read response rows from `pg_try_advisory_lock` function: %v", err)
+			return false, fmt.Errorf("error while trying to read response rows from locking function: %w", err)
 		}
-		return false, fmt.Errorf("missing response row from `pg_try_advisory_lock` function")
+		return false, fmt.Errorf("missing response row from locking function")
 	}
 	var result bool
```

Comment:
I think we should check rows.Err() after the scan to make sure no other errors came up

Reply:
I'm going to double check the pgx code, but I'm _pretty_ sure that the err is only populated on `Next()` and `Scan` is "pure" in the sense that it reads no new data from the DB and just tries to decode what it's already got

Reply:
The resolution was to switch to `QueryRow()` and avoid all issues